src/app/3d/qgs3dmaptoolpointcloudchangeattribute.cpp

```diff
@@ -203,50 +206,122 @@ SelectedPoints Qgs3DMapToolPointCloudChangeAttribute::searchPoints( QgsPointClou
+    const QgsPointCloudNodeId n( chunk->tileId().d, chunk->tileId().x, chunk->tileId().y, chunk->tileId().z );
+    const QVector<int> pts = selectedPointsInNode( searchPolygon, n, mapToPixel3D, layer );
     if ( !pts.isEmpty() )
     {
-      const QgsPointCloudNodeId n( chunk->tileId().d, chunk->tileId().x, chunk->tileId().y, chunk->tileId().z );
       result.insert( n, pts );
     }
   }
   return result;
 }
 
 
-QVector<int> Qgs3DMapToolPointCloudChangeAttribute::selectedPointsInNode( const QgsGeometry &searchPolygon, const QgsChunkNode *ch, const MapToPixel3D &mapToPixel3D, QgsPointCloudIndex &pcIndex )
+QVector<int> Qgs3DMapToolPointCloudChangeAttribute::selectedPointsInNode( const QgsGeometry &searchPolygon, const QgsPointCloudNodeId &n, const MapToPixel3D &mapToPixel3D, QgsPointCloudLayer *layer ) const
 {
   QVector<int> selected;
 
-  const QgsPointCloudNodeId n( ch->tileId().d, ch->tileId().x, ch->tileId().y, ch->tileId().z );
+  // Get the map's clipping extent in layer crs and skip if empty. We only need points within this extent.
+  const Qgs3DMapSettings *map = mCanvas->mapSettings();
+  const QgsRectangle mapExtent = Qgs3DUtils::tryReprojectExtent2D( map->extent(), map->crs(), layer->crs(), map->transformContext() );
+  if ( mapExtent.isEmpty() )
+    return selected;
+
+  QgsPointCloudIndex pcIndex = layer->index();
+  QgsPointCloudAttributeCollection attributes;
+  attributes.push_back( QgsPointCloudAttribute( QStringLiteral( "X" ), QgsPointCloudAttribute::Int32 ) );
+  attributes.push_back( QgsPointCloudAttribute( QStringLiteral( "Y" ), QgsPointCloudAttribute::Int32 ) );
+  attributes.push_back( QgsPointCloudAttribute( QStringLiteral( "Z" ), QgsPointCloudAttribute::Int32 ) );
+
+  QString categoryAttributeName;
+  QSet<int> categoryValues;
+  if ( QgsPointCloudLayer3DRenderer *renderer = dynamic_cast<QgsPointCloudLayer3DRenderer *>( layer->renderer3D() ) )
+  {
+    if ( const QgsClassificationPointCloud3DSymbol *symbol = dynamic_cast<const QgsClassificationPointCloud3DSymbol *>( renderer->symbol() ) )
+    {
+      const QgsPointCloudCategoryList categories = symbol->categoriesList();
+      for ( const auto &category : categories )
+      {
+        if ( category.renderState() )
+          categoryValues.insert( category.value() );
+      }
+
+      const QgsPointCloudAttributeCollection allAttributes = pcIndex.attributes();
+      categoryAttributeName = symbol->attribute();
+      attributes.extend( allAttributes, { categoryAttributeName } );
+    }
+  }
+
+  // we also need the filter expression so we can exclude filtered out points
+  QgsPointCloudExpression filterExpression = pcIndex.subsetString();
+  attributes.extend( pcIndex.attributes(), filterExpression.referencedAttributes() );
+
   QgsPointCloudRequest request;
-  // TODO: apply filtering (if any)
-  request.setAttributes( pcIndex.attributes() );
+  request.setAttributes( attributes );
+  request.setIgnoreIndexFilterEnabled( true );
```

Comment:
could you please add a comment why we are ignoring the index filter? to get all points so we have the correct point indexes within the node (and it could also go to docstring for setIgnoreIndexFilterEnabled() to clarify when one would want to use it)